.upstream-tests/test/support/cuda_space_selector.h

```diff
@@ -32,6 +32,19 @@
 #define SHARED
 #endif
 
+template<class Fun>
+__host__ __device__
+void execute_on_main_thread(Fun&& fun) {
```

Comment:
I'm guessing we'll need to add `__exec_check_disable__` to this, otherwise nvcc will throw warnings about invoking host-only functions in a host/device function.

Also, I think we need to use a wrapper macro for `__host__ __device__` to omit those attributes when compiling with a host-only compiler. Otherwise, gcc would complain about not knowing what these symbols mean. 

Usually looks something like this:
```
#ifdef __CUDACC__
#define __LIBCUDACXX_HOST_DEVICE __host__ __device__
#else
#define __LIBCUDACXX_HOST_DEVICE
#endif
```

Reply:
Oh those tests are only compiled with a cuda compiler. 

Only the libcxx tests are compiled with a pure host compiler

Reply:
Also the only purpose of that function is to dispatch between host and device, so that function passed better be `__host__ __device__`

Reply:
> Also the only purpose of that function is to dispatch between host and device, so that function passed better be __host__ __device__

It won't be. For example:
```
    execute_on_main_thread([&]{
        t = new cuda::std::atomic_flag();
        cuda::std::atomic_flag_clear(t);
        cuda::std::atomic_flag_wait(t, true);
    });
```

When `fake_main` runs on the host, this will be an implicitly host-only lambda.

When `fake_main` runs on the device, this will be a lambda defined inside a kernel (so no `__device__` annotation is needed).

https://godbolt.org/z/T9vPbPraf

Reply:
Added the exec check disable annotation